### ISSUE TITLE
host.docker.internal doesn't resolve on Linux

### ISSUE THREAD

**Issue by matthieuml** [docker/genai-stack#39]
As of my version of Docker:
```
Client: Docker Engine - Community
 Version:           24.0.6
 API version:       1.43
 Go version:        go1.20.7
 Git commit:        ed223bc
 Built:             Mon Sep  4 12:31:44 2023
 OS/Arch:           linux/amd64
 Context:           default
```

when launching the stack with ```docker compose --profile linux up```, the `pull-model` service fails to resolve host.docker.internal:
```
genai-stack-pull-model-1  | pulling ollama model llama2 using http://host.docker.internal:11434
genai-stack-pull-model-1  | Error: Head "http://host.docker.internal:11434/": dial tcp: lookup host.docker.internal on 127.0.0.11:53: no such host
```

There are multiple solutions:
1) Add in the docker-compose for `pull-model` and each `bot`:
```
extra_hosts:
      - "host.docker.internal:host-gateway"
 ```
2) We can just the change ```OLLAMA_BASE_URL=http://llm:11434```, but then we don't need the port mapping:
```
ports:
      - 11434:11434
```
inside of the `llm` service.

I feel like the second option feels more natural, should the port mapping then be removed to avoid confusion as the service is targeted for Linux ?

**Comment by karthik-mukund** [comment]
Works like a charm! Thank you for the timely response.

**Comment by andrealorenzon** [comment]
On my machine, your first solution worked by changing your error in this error:

```
genai-stack-pull-model-1  | ... pulling model (210s) - will take several minutes
genai-stack-pull-model-1  | ... pulling model (220s) - will take several minutes
genai-stack-pull-model-1  | ... pulling model (230s) - will take several minutes
genai-stack-pull-model-1  | ... pulling model (240s) - will take several minutes
genai-stack-pull-model-1  | Error: Head "http://host.docker.internal:11434/": dial tcp 172.17.0.1:11434: connect: connection timed out
genai-stack-pull-model-1 exited with code 1
```

What is that dial referred to?

**Comment by matthieuml** [comment]
Is this working now ? It looks like you can't access Neo4j either because it is not running, or a network related issue.

Maybe you should open an issue about it, it should be easier for other to see it!

**Comment by andrealorenzon** [comment]
deleting `/data` content solved the issue.

**Comment by WindyOrgUK** [comment]
Any chance for a full working docker-compose.yml for linux and linux-gpu ?
None of the 2 options worked for me (not sure I implemented them correctly)

**Comment by matthieuml** [comment]
The current docker-compose fails in your case?

**Comment by WindyOrgUK** [comment]
cloned the repo a few hours ago.

**Comment by WindyOrgUK** [comment]
Please ignore me.
I did not edit the .env file before.
The problem got fixed as soon as I uncommented the:
OLLAMA_BASE_URL=http://host.docker.internal:11434

**Comment by WindyOrgUK** [comment]
deleted my previous comment.
The problem was not fixed actually by editing the .env file in the section I mentioned.
I just switched the llm from llama to openai which made connection to port 11434 not needed anymore

**Comment by matthieuml** [comment]
Docker doesn't recognize `host.docker.internal`, also, you would need to modify the `/etc/hosts` inside of the container for it to take effect.

As I have written at the beginning of this issue, you can try to add this in the `docker-compose`, it is the equivalent of what you are trying to do
```
extra_hosts:
      - "host.docker.internal:host-gateway"
```

**Comment by WindyOrgUK** [comment]
I believe I tested earlier all the previously mentioned solutions.
Which exactly sections of the docker-compose.yml file each of these changes need to be made ?

**Comment by matthieuml** [comment]
What is the stacktrace of the error when using http://llm:11434 in the `.env`?.

**Comment by WindyOrgUK** [comment]
Just a correction, I kept trying linux-gpu which was giving me this error.
with "linux" it just now worked.
Is there a solution for "linux-gpu" ?

**Comment by matthieuml** [comment]
Oh I see, when you use service `linux-gpu`, the new address to put in `.env` is http://linux-gpu:11434!

**Comment by WindyOrgUK** [comment]
Perfect, that worked too.
Thx a lot for your help

**Comment by fahadaziz52090** [comment]
I am struggling with this same error . I have attached my env and  yml file. 
when I go to localhost:11434 it says ollama is running. 

I have tried all of above solution. I am new to this . 

Error Says :

pulling ollama model llama2 using http://host.docker.internal:11434
pull-model-1  | Error: Head "http://host.docker.internal:11434/": dial tcp: lookup host.docker.internal on 127.0.0.11:53: no such host
pull-model-1 exited with code 1


[env.txt](https://github.com/docker/genai-stack/files/14005631/env.txt)
[docker-compose.txt](https://github.com/docker/genai-stack/files/14005632/docker-compose.txt)
